clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+     * \p module_mapper_t.
+     *
+     * If there is a decoding failure, the default-constructed DecodeInfo that
+     * returns is_valid() = false will be added to the cache.
+     *
+     * Returns the error string, or an empty string if there was no error.
+     */
+    std::string
+    add_decode_info(const dynamorio::drmemtrace::_memref_instr_t &memref_instr)
+    {
+        if (!init_done_) {
+            return "init() must be called first";
+        }
+        const app_pc trace_pc = reinterpret_cast<app_pc>(memref_instr.addr);
+        if (!use_module_mapper_ && memref_instr.encoding_is_new) {
+            // If the user asked to use the module mapper despite having
+            // embedded encodings in the trace, we don't invalidate the
+            // cached encoding.
+            decode_cache_.erase(trace_pc);
+        }
+        auto it = decode_cache_.find(trace_pc);
+        // If the prior cached info was invalid, we try again.
+        if (it != decode_cache_.end() && it->second.is_valid()) {
+            return "";
+        }
+        decode_cache_[trace_pc] = DecodeInfo();
+        instr_t *instr = nullptr;
+        instr_noalloc_t noalloc;
+        if (persist_decoded_instrs_) {
+            instr = instr_create(dcontext_);
+        } else {
+            instr_noalloc_init(dcontext_, &noalloc);
+            instr = instr_from_noalloc(&noalloc);
+        }
+
+        app_pc decode_pc;
+        if (!use_module_mapper_) {
+            decode_pc = const_cast<app_pc>(memref_instr.encoding);
+        } else {
+            // Legacy trace support where we need the binaries.
+            std::string err = find_mapped_trace_address(trace_pc, decode_pc);
+            if (err != "")
+                return err;
+        }
+        app_pc next_pc = decode_from_copy(dcontext_, decode_pc, trace_pc, instr);
+        if (next_pc == nullptr || !instr_valid(instr)) {
+            if (persist_decoded_instrs_) {
+                instr_destroy(dcontext_, instr);
+            }
+            return "decode_from_copy failed";
+        }
+        decode_cache_[trace_pc].set_decode_info(dcontext_, memref_instr, instr);
+        return "";
+    }
+
+    /**
+     * Performs initialization tasks such as verifying whether the given trace
+     * indeed has embedded encodings or not, and initializing the
+     * \p module_mapper_t if the module path is provided.
+     *
+     * It is important to note that the trace filetype may be obtained using the
+     * get_filetype() API on a \p memtrace_stream_t. However, not all instances of
+     * \p memtrace_stream_t have the filetype available at init time (before the
+     * \p TRACE_MARKER_TYPE_FILETYPE is actually returned by the stream).
+     * Particularly, the \p stream_t provided by the \p scheduler_t to analysis
+     * tools provides the file type at init time when #dynamorio::drmemtrace::
+     * scheduler_tmpl_t::scheduler_options_t.read_inputs_in_init is set. In other
```

Comment:
I think this is too much detail breaking abstraction barriers without providing key info the user needs to know.  Analysis tools don't need to know about scheduler paramters; they have have no control over that.  Just say that the filetype is available in offline mode but not for online analysis.

Reply:
Done